classes/manager.php

```diff
@@ -163,12 +168,190 @@ public static function get_reasons(float $duration): int {
             $reason |= self::REASON_FLAMEALL;
         }
-        if (($duration * 1000) >= (int) get_config('tool_excimer', 'trigger_ms')) {
+
+        if (($duration * 1000) >= self::get_min_duration_for_reason_slow()) {
             $reason |= self::REASON_AUTO;
         }
         return $reason;
     }
 
+    /**
+     * Returns the minimum duration for profiles matching this reason and page/request.
+     *
+     * @param  int $reason - the profile type or REASON_*
+     * @return float duration (in milliseconds) of the fastest profile for a given reason and request/page.
+     */
+    public static function get_min_duration_for_request_and_reason(string $request, int $reason): float {
+        global $DB;
+
+        $reasonstr = self::REASON_STR_MAP[$reason];
+        $pagequota = (int) get_config('tool_excimer', "num_' . $reasonstr . '_by_page");
+
+        // Grab the fastest profile for this page/request, and use that as
+        // the lower boundary for any new profiles of this page/request.
+        $cachekey = "profile_type_$reason" . "_page_$request" . '_min_duration_ms';
+        $cache = \cache::make('tool_excimer', 'timings');
+        $result = $cache->get($cachekey);
+        if ($result === false) {
+            // NOTE: Opting to query this way instead of using MIN due to
+            // the fact valid profiles will be added and the limits will be
+            // breached for 'some time'. This will keep the constraints as
+            // correct as possible.
+            $reasons = $DB->sql_bitand('reason', $reason);
+            $sql = "SELECT duration as min_duration
+                      FROM {tool_excimer_profiles}
+                     WHERE $reasons != ?
+                           AND request = ?
+                  ORDER BY duration DESC
+                     ";
+            $resultset = $DB->get_records_sql($sql, [
+                self::REASON_NONE,
+                $request,
+            ], 0, $pagequota);
+            // Cache the results in milliseconds (avoids recalculation later).
+            $result = (end($resultset)->min_duration ?? 0) * 1000;
+            $cache->set($cachekey, $result);
+        }
+        return $result;
+    }
+
+    /**
+     * Returns the minimum duration for profiles matching this reason.
+     *
+     * @param  int $reason - the profile type or REASON_*
+     * @return float duration (in milliseconds) of the fastest profile for a given reason.
+     */
+    public static function get_min_duration_for_reason(int $reason): float {
+        global $DB;
+
+        $reasonstr = self::REASON_STR_MAP[$reason];
+        $quota = (int) get_config('tool_excimer', "num_$reasonstr");
+        $cache = \cache::make('tool_excimer', 'timings');
+        // Grab the fastest profile across the slow profiles, and use that
+        // as the lower boundary for any new profiles.
+        $cachekey = "profile_type_$reason" . '_min_duration_ms';
+        $result = $cache->get($cachekey);
```

Comment:
This should only be a single number across the system (per reason), so I would not use the cache api this, I'd store it directly using get / set_config and then it is simply there already without another cache read as get_config loads all config for a plugin in one call.